src/bootstrap/bootstrap.py

```
@@ -1288,6 +1294,13 @@ def bootstrap(args):
         with open(include_path) as included_toml:
             config_toml += os.linesep + included_toml.read()
 
+    if is_non_git_source:
+        for option in ["download-ci-llvm", "download-rustc"]:
+            if RustBuild.get_toml_static(config_toml, option):
+                raise Exception(
+                    "Option '{}' cannot be used with non-git sources.".format(option)
+                )
+
```

I think all this logic could be implemented directly in the rust-side bootstrap. We want to keep the python part as minimal as possible since we plan to migrate all python code to rust.

---

@onur-ozkan, thanks for reviewing this PR!

Could you please guide me to the correct file locations in [rust-side bootstrap](https://github.com/rust-lang/rust/blob/master/src/bootstrap) where I should implement this logic (and the tests)?  
I checked the codebase and I think the logic should be implemented in [src/core/config/config.rs](https://github.com/rust-lang/rust/blob/bd62a452f45c2386275c5469f9ae390e7de2dc6f/src/bootstrap/src/core/config/config.rs) and tests should go in [src/core/config/tests.rs](https://github.com/rust-lang/rust/blob/bd62a452f45c2386275c5469f9ae390e7de2dc6f/src/bootstrap/src/core/config/tests.rs) in [bootstrap](https://github.com/rust-lang/rust/tree/bd62a452f45c2386275c5469f9ae390e7de2dc6f/src/bootstrap). I'm not sure though.

---

Yes, these are the correct places. You can see it's already done for ci-rustc:

https://github.com/rust-lang/rust/blob/c62b732724b6beb50a5c73be26d0c5f668504059/src/bootstrap/src/core/config/config.rs#L2852-L2858

so you only have to handle it for ci-llvm.

---

In fact, we also have the coverage for ci-llvm, but it's silently fallbacking to disable ci-llvm instead of hard-error like ci-rustc

https://github.com/rust-lang/rust/blob/c62b732724b6beb50a5c73be26d0c5f668504059/src/bootstrap/src/core/config/config.rs#L2927-L2932

you can modify that to follow same approach as ci-rustc (doing the hard-error).